_posts/2018-10-04-js-testing.md

```diff
@@ -0,0 +1,206 @@
+---
+layout: post
+title:  "Javascript Projects Testing"
+date:   2018-10-04 12:00:00 -0400
+categories: software testing js javascript jest enzyme
+author: Tania Paiva
+---
+
+Should I test this? the answer is a BIG YES. Tests are your friends, and saviors sometimes (often).
+
+At Forward Financing, we feel proud when our projects reach high test coverage. It's our best insurance  
+for making improvements and upgrades later with the warranty that existing functionality will  
+keep working.
+
+Mistakes can always happen, and that is where tests can be real saviors, as an alert that  
+something was altered by mistake, or that some functionality is not providing the expected result.
+
+Tests verify the application meets the requeriments, it guarantees the code makes what we expect  
+it to do.
+
+Okay, tests can do a lot for us, but is not always sun and butterflies, writing good tests takes  
+time and it gets better with some experience.
+
+![Coverage]({{"/assets/test_coverage.png" | absolute_url}})
+
+In this article we will talk about the testing tools we use for our javascript projects.
+
+* [Jest](https://jestjs.io/)
+Delightful JavaScript Testing, Jest is used by Facebook to test all JavaScript code including  
+React applications. One of Jest's philosophies is to provide an integrated "zero-configuration" experience. 
+
+* [Enzyme](https://airbnb.io/enzyme/)
+Is a JavaScript Testing utility for React that makes it easier to assert, manipulate, and traverse your  
+React Components' output.
+
+Ok, those are cool names but what can we do with them?  
+Here are some examples on how to use these tools:
+
+We have a utility function that validates dates are in the following format DD/MM/YYYY
+
+``` javascript
+function areInvalidDates(values){
+    let status = [];
+    if (values.length > 0) {
+      status = values.map(value => {
+        if (value.split("/").length !== 3 ||
+            value.split("/")[2].length !== 4 ||
+            value.split("/")[1].length !== 2 ||
+            value.split("/")[0].length !== 2
+            ) {
+            return 'invalid';
+        } else {
+          return 'valid';
+        }
+      });
+    }
+    return status.includes('invalid');
```

Comment:
Let's see if we can shorten this code just a little. I don't think there's any real reason to be collecting an array on strings (really booleans, just named `valid` and `invalid`). In general, if you see yourself returning a boolean, and you have a list/array, you might want `andmap`, or `ormap`, or (if you're language doesn't support it) the even more general `fold(l/r)`, a.k.a. `reduce` in JS.

In this case however, you could spare the map altogether and just loop over the `values` directly, and `return` from the whole function when you find the first `invalid`. Personally I'd make this only validate a single date, then `andmap` it in the tests on the array of dates to be tested.

Reply:
I used this example because is a real test we have in our code, but we can change it.